Implemented Response as an object instead of a hash

Sacristan opened the pull request:
* Moved  ActiveMerchant::Billing::FirstDataGateway to ActiveMerchant::Billing::FirstData::Gateway
* Added ActiveMerchant::Billing::FirstData::Response -> now response is not just a hash, but object from what fields like transaction_id, parameters can be returned, also contains ok?, recurring? helper methods;
* Added result_text method to ActiveMerchant::Billing::FirstData::Response instance which returns description of result (not just code or OK, but an explanation to the status code);
* Updated Rspec to v 3.2;
* Rspec tests changed from should to expect style ;
* Added specs for ActiveMerchant::Billing::FirstData::Response ;
* ActiveMerchant::Billing::FirstData::Gateway.ssl_strict is to be set directly

ebeigarts commented:
This is great :)

Can you make this backwards compatible by using  `FirstDataGateway` and `FirstDataResponse` classes and adding `Response#[]` method?

Also this kind of naming convention is used by `active_merchant` gem - `<name>Gateway` and `<name>Response`.

It would be nice if `FirstDataResponse` would inherit from [Response](https://github.com/Shopify/active_merchant/blob/master/lib/active_merchant/billing/response.rb) and implement `#success?`, `#authorization` (transaction_id), etc methods.

Global constants are bad, can you move them to `ActiveMerchant::Billing::FirstDataGateway::CURRENCY_CODES` and `ActiveMerchant::Billing::FirstDataResponse::FDL_RESPONSE_CODES`?

Sacristan commented:
sure - ill update it ASAP

Sacristan commented:
Im sorry - what You meant with "adding Response#[] method" ?

ebeigarts commented:
```
class Response
  def [](value)
    @parameters[value]
  end
end
```

so we can access the values the old way using `response[:transaction_id]`

Sacristan commented:
Got it! Thanks!

Sacristan commented:
"
It would be nice if FirstDataResponse would inherit from Response and implement #success?, #authorization (transaction_id), etc methods.
"

authorization -> transaction_id ?

Current methods for ResponseObject
ok?
success? <- alias for ok?
recurring?
result
transaction_id
result_code
recc_pmnt_id
recc_pmnt_expiry
three_dee_secure
result_text
parameters

Sacristan commented:
* Constants moved to their representable classes;
* Backwards compatibility ensured for Response and Gateway classes;
* Moved self.ssl_strict = false as it was in 1.2.0 (not configurable);
* Added authorization, success?, [], message, result_message methods to response.

ebeigarts commented:
I think `FirstData::Gateway` is not needed if we have `FirstDataGateway`, can you move all the methods back?

ebeigarts commented:
More comments :)

The `c` prefix is confusing in `c000: "Approved"`. `'000' => 'Approved'` would be better and less confusing.

I would prefer `Response#_3d_secure` over `#three_dee_secure`.

Why do you need that `FDL_` prefix? (`FDL_RESPONSE_CODES` => `RESPONSE_CODES`)

Since we now have `Response#success?`, we could remove `Response#ok?`.

Why do we need `Response#[]`? it should enough with `Response#[](value)`.

Inheriting `FirstDataResponse` from `ActiveMerchant::Billing::Response` would allow us
to reuse some functionality and be compatible with `ActiveMerchant::Billing::Response`.
There could be people that are using multiple gateways and expect gateways to return responses that respond to the exact same methods.

```ruby
class FirstDataResponse < Response
  def initialize(params, options = {})
    success == params[:result] == 'OK'
    message = FDL_RESPONSE_CODES[options[:result_code]]
    error_code = options[:result_code] unless success
    super success, message, params, options.merge(
      authorization: params[:transaction_id]
    )
  end

  [:transaction_id, :result_code, ...].each do |name|
    define_method name do
      params[name]
    end
  end

  ...
end
```

Sacristan commented:
Please evaluate the code once again!

ebeigarts commented on lib/active_merchant/billing/first_data_gateway.rb:
```diff
@@ -286,34 +284,33 @@ def redirect_url(trans_id = nil)
 
-      def lookup_currency(params)
-        params[:currency] = CURRENCY_CODES[params[:currency]] || params[:currency]
-      end
```
Can you leave the indention style as it was previously?

ebeigarts commented on lib/active_merchant/billing/first_data_response.rb:
```diff
@@ -0,0 +1,154 @@
+          authorization: params[:transaction_id]
+        )
+
+        self.parameters=params
```
why do you need `self.parameters` if we already have `self.params`?

ebeigarts commented on spec/active_merchant/billing/first_data/gateway_spec.rb:
```diff
@@ -42,130 +43,146 @@
+    it "should be present" do
+      expect(@gateway.present?).to be true
+    end
+  end
```
not sure this test is required anymore

ebeigarts commented on spec/active_merchant/billing/first_data/gateway_spec.rb:
```diff
@@ -230,75 +247,75 @@
-        response[:result_code].should == "000"
+        puts "Response"
+        puts response.inspect
+        puts response.class
```
No puts please :)

ebeigarts commented on lib/active_merchant/billing/first_data_response.rb:
```diff
@@ -0,0 +1,154 @@
+      end
+
+      def result_text
+        RESPONSE_CODES[result_code]
```
it is already set in initialzier as `message`

```
def result_text
  message
end
```

ebeigarts commented:
Also tests are failing :(

Sacristan commented:
@ebeigarts im back into this. Ill try fixing all the problems and then merge in Your master changes

Sacristan commented:
Merged in master. Please evaluate code

ebeigarts commented:
@Sacristan thnx, merged

Sacristan commented:
@ebeigarts Can You make a release ?

ebeigarts commented:
Released v1.3.0, let me know it there are issues.

> On 13 Oct 2015, at 20:09, Ģirts Ķesteris <notifications@github.com> wrote:
> 
> @ebeigarts <https://github.com/ebeigarts> Can You make a release ?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/ebeigarts/active_merchant_first_data/pull/6#issuecomment-147780324>.
>
